File: scripts/state_archive.py
"""Vela state/2 HDF5 contract. No legacy file sniffing or format fallback."""
import json
import math
import os
from pathlib import Path
import re
import tempfile
import hashlib
import struct

import h5py
import numpy as np

SCHEMA = 'vela.state/2'
MAX_BYTES = 2**30
UNITS = {name:'V' for name in ('psi', 'phin', 'phip', 'electron_quantum_potential_V',
    'electron_quantum_potential_like_V', 'electron_qf_increment_V', 'hole_qf_increment_V',
    'electron_qf_reference_V', 'hole_qf_reference_V')}
UNITS.update(electrons_m3='m^-3', holes_m3='m^-3', temperature_K='K')
REQUIRED = {'psi', 'phin', 'phip', 'electrons_m3', 'holes_m3'}
SPLIT = {'electron_qf_increment_V', 'hole_qf_increment_V',
         'electron_qf_reference_V', 'hole_qf_reference_V'}


def mesh_identity(mesh, length_m_per_internal):
    """Canonical encoding shared with stateMeshIdentity, independent of JSON spacing."""
    if not math.isfinite(length_m_per_internal) or length_m_per_internal <= 0:
        raise ValueError('State mesh length unit must be positive and finite')
    data = bytearray(b'vela.mesh/1')
    def integer(v): data.extend(struct.pack('<Q', v))
    def real(v): data.extend(struct.pack('<d', v))
    def text(v):
        b = v.encode('utf-8'); integer(len(b)); data.extend(b)
    def indices(values):
        integer(len(values))
        for v in values: integer(v)
    real(length_m_per_internal)
    integer(len(mesh['nodes']))
    for n in mesh['nodes']: integer(n['id']); real(n['x']); real(n['y'])
    integer(len(mesh['triangles']))
    for c in mesh['triangles']: integer(c['id']); integer(c['region_id']); indices(c['node_ids'])
    integer(len(mesh['regions']))
    for r in mesh['regions']: integer(r['id']); text(r['name']); text(r['material']); indices(r['cell_ids'])
    integer(len(mesh['contacts']))
    for c in mesh['contacts']:
        integer(c['id']); text(c['name']); integer(c['region_id']); indices(c['node_ids'])
        edges = c.get('edge_node_ids', [])
        integer(len(edges))
        for a,b in edges: integer(a); integer(b)
    return hashlib.sha256(data).hexdigest()


def fields_to_rows(fields):
    # Keep binary64 values numeric. Formatting every element as decimal here
    # only makes the predictor parse it again; CSV export owns text formatting.
    names=list(fields)
    columns=[np.asarray(fields[name],dtype='<f8').tolist() for name in names]
    return [dict(zip(['node_id']+names,[str(i),*values])) for i,values in enumerate(zip(*columns))]


def inspect(path):
    """Read a self-described archive for reporting, without qualifying its mesh.

    Solver/seed callers must use read() with an independently known mesh identity.
    This entry is for table inspectors and conversion reports only.
    """
    path = Path(path)
    if path.suffix != '.h5' or path.stat().st_size > MAX_BYTES + 1024*1024:
        raise ValueError('Invalid state archive path/size')
    with h5py.File(path,'r') as f:
        metadata = json.loads(f.attrs['metadata_json'])
        count = int(f.attrs['node_count'])
    return read(path,count,metadata['mesh_sha256'])


def read_rows(path):
    return fields_to_rows(inspect(path)[0])


def rows_to_fields(rows):
    if not rows: raise ValueError('Empty state')
    names = set(rows[0])-{'node_id'}
    if not REQUIRED <= names or not names <= UNITS.keys(): raise ValueError('Unknown state columns')
    fields = {name:np.empty(len(rows),dtype='<f8') for name in names}
    seen = set()
    for row in rows:
        node = int(row['node_id'])
        if node in seen or node < 0 or node >= len(rows) or set(row)-{'node_id'} != names:
            raise ValueError('Invalid state node order or columns')
        seen.add(node)
        for name in names: fields[name][node] = float(row[name])
    return fields


def translate(source, destination, subtract, active_nodes, expected_nodes, expected_mesh_sha256):
    if not math.isfinite(subtract): raise ValueError('Invalid frame offset')
    if Path(destination).exists(): raise FileExistsError(destination)
    fields, metadata = read(source, expected_nodes, expected_mesh_sha256)
    active = set(active_nodes)
    for i in range(expected_nodes):
        live = i in active
        for carrier in ('electron','hole'):
            ref,inc = carrier+'_qf_reference_V',carrier+'_qf_increment_V'
            if ref in fields:
                a,b=float(fields[ref][i]),float(fields[inc][i]);new=a-subtract
                fields[ref][i]=new if live else 0.
                fields[inc][i]=math.fsum((a,-subtract,-new,b)) if live else 0.
        for key in ('psi','phin','phip'):
            fields[key][i]=float(fields[key][i])-subtract if key=='psi' or live else 0.
    metadata['potential_origin_V'] += subtract
    # Saved DD bias values are expressed in the solver reference frame.
    # Physical terminal bias = stored bias + potential_origin_V.
    if 'bias_V' in metadata: metadata['bias_V'] -= subtract
    if 'contact_biases_V' in metadata:
        metadata['contact_biases_V']={k:v-subtract for k,v in metadata['contact_biases_V'].items()}
    write(destination, fields, metadata)


def validate(fields, metadata):
    required = REQUIRED if metadata.get('mode') == 'dd' else {'psi', 'phin', 'phip', 'temperature_K'}
    if not required <= fields.keys() or not fields.keys() <= UNITS.keys():
        raise ValueError('Missing or unknown state field')
    if ('electrons_m3' in fields) != ('holes_m3' in fields):
        raise ValueError('Partial density pair')
    count = len(fields['psi'])
    if count < 1 or count * len(fields) * 8 > MAX_BYTES:
        raise ValueError('Invalid or oversized dimensions')
    if metadata.get('mode') not in ('dd', 'electrothermal'):
        raise ValueError('Unknown equation mode')
    if not re.fullmatch('[0-9a-f]{64}', metadata.get('mesh_sha256', '')):
        raise ValueError('Mesh identity required')
    if type(metadata['potential_origin_V']) not in (int, float) or not math.isfinite(metadata['potential_origin_V']):
        raise ValueError('Invalid potential origin')
    if len(json.dumps(metadata, allow_nan=False).encode()) > 65536:
        raise ValueError('Metadata too large')
    if ('temperature_K' in fields) != (metadata['mode'] == 'electrothermal'):
        raise ValueError('Temperature/mode mismatch')
    if 'electron_quantum_potential_like_V' in fields and 'electron_quantum_potential_V' not in fields:
        raise ValueError('Partial quantum state')
    if len(SPLIT.intersection(fields)) not in (0, 4):
        raise ValueError('Partial split QF state')
    for name, values in fields.items():
        a = np.asarray(values, dtype='<f8')
        if a.shape != (count,) or not np.all(np.isfinite(a)):
            raise ValueError('Invalid field '+name)
    if SPLIT <= fields.keys():
        for carrier, qf in (('electron', 'phin'), ('hole', 'phip')):
            combined = np.asarray(fields[carrier+'_qf_reference_V']) + np.asarray(fields[carrier+'_qf_increment_V'])
            value = np.asarray(fields[qf])
            ceiling = 32*np.finfo(float).eps*np.maximum(1., np.maximum(abs(value), abs(combined)))
            if not np.all(np.isfinite(combined)) or np.any(abs(value-combined) > ceiling):
                raise ValueError('Inconsistent split QF state')
    return count


def write(path, fields, metadata):
    path = Path(path)
    if path.suffix != '.h5':
        raise ValueError('State path must end in .h5')
    count = validate(fields, metadata)
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, temporary = tempfile.mkstemp(prefix=path.name+'.tmp.', dir=path.parent)
    os.close(fd)
    try:
        with h5py.File(temporary, 'w') as f:
            f.attrs['schema'] = SCHEMA
            f.attrs['node_count'] = np.uint64(count)
            f.attrs['metadata_json'] = json.dumps(metadata, allow_nan=False, separators=(',', ':'))
            group = f.create_group('fields')
            for name in sorted(fields):
                d = group.create_dataset(name, data=np.asarray(fields[name], dtype='<f8'))
                d.attrs['unit'] = UNITS[name]
            f.flush()
        os.replace(temporary, path)
    finally:
        if os.path.exists(temporary):
            os.unlink(temporary)


def read(path, expected_nodes, expected_mesh_sha256):
    path = Path(path)
    if path.suffix != '.h5' or path.stat().st_size > MAX_BYTES+2**20:
        raise ValueError('Invalid path or oversized state')
    if not re.fullmatch('[0-9a-f]{64}', expected_mesh_sha256):
        raise ValueError('Expected mesh identity required')
    def text(value):
        if isinstance(value, bytes): return value.decode('utf-8')
        if not isinstance(value, str): raise ValueError('Expected scalar string attribute')
        return value
    with h5py.File(path, 'r') as f:
        if text(f.attrs['schema']) != SCHEMA:
            raise ValueError('Unsupported schema')
        raw_count = np.asarray(f.attrs['node_count'])
        if raw_count.shape != () or raw_count.dtype.kind != 'u' or raw_count.dtype.itemsize != 8:
            raise ValueError('Node count must be scalar uint64')
        count = int(raw_count)
        if count != expected_nodes or count < 1:
            raise ValueError('Node count mismatch')
        raw = text(f.attrs['metadata_json'])
        if len(raw.encode()) > 65536: raise ValueError('Metadata too large')
        metadata = json.loads(raw)
        if metadata['mesh_sha256'] != expected_mesh_sha256:
            raise ValueError('Mesh identity mismatch')
        group = f['fields']
        if not 4 <= len(group) <= len(UNITS) or count*len(group)*8 > MAX_BYTES:
            raise ValueError('Invalid or oversized dimensions')
        fields = {}
        for name in group:
            if name not in UNITS: raise ValueError('Unknown field')
            d = group[name]
            if d.shape != (count,) or d.dtype != np.dtype('<f8') or text(d.attrs['unit']) != UNITS[name]:
                raise ValueError('Field type, dimensions or units mismatch')
            fields[name] = d[...]
    validate(fields, metadata)
    return fields, metadata

File: src/io/StateArchive.cpp
#include "vela/io/StateArchive.h"
#include <highfive/highfive.hpp>
#include <algorithm>
#include <atomic>
#include <chrono>
#include <cmath>
#include <limits>
#include <set>
#include <stdexcept>
#ifdef _WIN32
#ifndef NOMINMAX
#define NOMINMAX
#endif
#include <windows.h>
#endif

namespace vela {
namespace {
constexpr const char* schema = "vela.state/2";
constexpr std::uint64_t maxBytes = 1024ULL * 1024 * 1024;
const std::map<std::string, std::string> units{
    {"psi", "V"}, {"phin", "V"}, {"phip", "V"},
    {"electrons_m3", "m^-3"}, {"holes_m3", "m^-3"},
    {"electron_quantum_potential_V", "V"}, {"electron_quantum_potential_like_V", "V"},
    {"electron_qf_increment_V", "V"}, {"hole_qf_increment_V", "V"},
    {"electron_qf_reference_V", "V"}, {"hole_qf_reference_V", "V"},
    {"temperature_K", "K"}};
[[noreturn]] void bad(const std::string& message) {
    throw std::runtime_error("HDF5 state: " + message);
}
bool hashValid(const std::string& hash) {
    return hash.size() == 64 && std::all_of(hash.begin(), hash.end(), [](char c) {
        return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f');
    });
}
void checkPath(const std::filesystem::path& path) {
    if (path.extension() != ".h5") bad("state path must end in .h5");
}
void dimensions(std::uint64_t count, std::uint64_t fields) {
    if (count == 0 || fields < 4 || fields > units.size() || count > maxBytes / (8 * fields))
        bad("invalid or oversized dimensions");
}
std::string textAttribute(const HighFive::File& file, const char* name) {
    const auto a = file.getAttribute(name);
    if (a.getSpace().getElementCount() != 1) bad(std::string("non-scalar attribute ") + name);
    return a.read<std::string>();
}
} // namespace

void validateStateArchive(const StateArchive& s) {
    dimensions(s.nodeCount, s.fields.size());
    if (!s.metadata.is_object() || s.metadata.dump().size() > 65536) bad("invalid metadata");
    const auto mode = s.metadata.at("mode").get<std::string>();
    if (mode != "dd" && mode != "electrothermal") bad("unknown equation mode");
    if (!hashValid(s.metadata.at("mesh_sha256").get<std::string>())) bad("invalid mesh identity");
    if (!s.metadata.at("potential_origin_V").is_number() ||
        !std::isfinite(s.metadata.at("potential_origin_V").get<double>())) bad("invalid potential origin");
    for (const auto* name : {"psi", "phin", "phip"})
        if (!s.fields.contains(name)) bad(std::string("missing field ") + name);
    if (mode == "dd" && (!s.fields.contains("electrons_m3") || !s.fields.contains("holes_m3")))
        bad("DD densities required");
    if (s.fields.contains("electrons_m3") != s.fields.contains("holes_m3")) bad("partial density pair");
    if ((mode == "electrothermal") != s.fields.contains("temperature_K")) bad("temperature/mode mismatch");
    if (s.fields.contains("electron_quantum_potential_like_V") &&
        !s.fields.contains("electron_quantum_potential_V")) bad("partial quantum state");
    std::size_t split = 0;
    for (const auto* name : {"electron_qf_increment_V", "hole_qf_increment_V",
                            "electron_qf_reference_V", "hole_qf_reference_V"})
        split += s.fields.contains(name);
    if (split != 0 && split != 4) bad("partial split quasi-Fermi state");
    for (const auto& [name, values] : s.fields) {
        if (!units.contains(name) || values.size() != s.nodeCount) bad("unknown or incomplete field " + name);
        if (!std::all_of(values.begin(), values.end(), [](double v) { return std::isfinite(v); }))
            bad("non-finite field " + name);
    }
    if (split) for (std::size_t i = 0; i < s.nodeCount; ++i) {
        for (const auto& carrier : {std::string("electron"), std::string("hole")}) {
            const double value = s.fields.at(carrier == "electron" ? "phin" : "phip")[i];
            const double combined = s.fields.at(carrier + "_qf_reference_V")[i] +
                                    s.fields.at(carrier + "_qf_increment_V")[i];
            if (!std::isfinite(combined) || std::abs(value - combined) >
                32 * std::numeric_limits<double>::epsilon() * std::max({1., std::abs(value), std::abs(combined)}))
                bad("inconsistent split quasi-Fermi state");
        }
    }
}

void writeStateArchive(const std::filesystem::path& path, const StateArchive& state) {
    checkPath(path);
    validateStateArchive(state);
    if (!path.parent_path().empty()) std::filesystem::create_directories(path.parent_path());
    static std::atomic<std::uint64_t> sequence{0};
    auto temporary = path;
    temporary += ".tmp." + std::to_string(std::chrono::steady_clock::now().time_since_epoch().count()) +
                 "." + std::to_string(sequence.fetch_add(1));
    try {
        {
            HighFive::File f(temporary.string(), HighFive::File::Create | HighFive::File::Excl);
            f.createAttribute("schema", std::string(schema));
            f.createAttribute("node_count", state.nodeCount);
            f.createAttribute("metadata_json", state.metadata.dump());
            auto group = f.createGroup("fields");
            for (const auto& [name, values] : state.fields) {
                HighFive::AtomicType<double> diskType;
                if (H5Tset_order(diskType.getId(), H5T_ORDER_LE) < 0 ||
                    H5Tequal(diskType.getId(), H5T_IEEE_F64LE) <= 0) bad("binary64 representation unavailable");
                auto d = group.createDataSet(name, HighFive::DataSpace({values.size()}), diskType);
                d.createAttribute("unit", units.at(name));
                d.write_raw(values.data());
            }
            f.flush();
        }
#ifdef _WIN32
        if (!MoveFileExW(temporary.c_str(), path.c_str(), MOVEFILE_REPLACE_EXISTING | MOVEFILE_WRITE_THROUGH))
            bad("cannot replace checkpoint, Windows error " + std::to_string(GetLastError()));
#else
        std::filesystem::rename(temporary, path);
#endif
    } catch (...) {
        std::error_code ignored;
        std::filesystem::remove(temporary, ignored);
        throw;
    }
}

StateArchive readStateArchive(const std::filesystem::path& path, std::uint64_t expected,
                              const std::string& mesh) {
    checkPath(path);
    if (!hashValid(mesh)) bad("expected mesh identity is required");
    if (std::filesystem::file_size(path) > maxBytes + 1024 * 1024) bad("file exceeds size limit");
    HighFive::File f(path.string(), HighFive::File::ReadOnly);
    if (textAttribute(f, "schema") != schema) bad("unsupported schema");
    StateArchive s;
    const auto countAttribute = f.getAttribute("node_count");
    const auto countType = countAttribute.getDataType();
    if (countAttribute.getSpace().getElementCount() != 1 || H5Tget_class(countType.getId()) != H5T_INTEGER ||
        H5Tget_size(countType.getId()) != 8 || H5Tget_sign(countType.getId()) != H5T_SGN_NONE)
        bad("node count must be scalar uint64");
    s.nodeCount = countAttribute.read<std::uint64_t>();
    if (s.nodeCount != expected) bad("node count mismatch");
    const auto rawMetadata = textAttribute(f, "metadata_json");
    if (rawMetadata.size() > 65536) bad("metadata too large");
    s.metadata = nlohmann::json::parse(rawMetadata);
    if (s.metadata.at("mesh_sha256") != mesh) bad("mesh identity mismatch");
    auto group = f.getGroup("fields");
    const auto names = group.listObjectNames();
    dimensions(s.nodeCount, names.size());
    for (const auto& name : names) {
        if (!units.contains(name)) bad("unknown field " + name);
        auto d = group.getDataSet(name);
        const auto type = d.getDataType();
        if (d.getDimensions() != std::vector<std::size_t>{static_cast<std::size_t>(s.nodeCount)} ||
            H5Tequal(type.getId(), H5T_IEEE_F64LE) <= 0) bad("field type or dimensions mismatch");
        if (d.getAttribute("unit").read<std::string>() != units.at(name)) bad("field unit mismatch");
        auto& values = s.fields[name];
        values.resize(s.nodeCount);
        d.read_raw(values.data());
    }
    validateStateArchive(s);
    return s;
}
} // namespace vela
